### PR TITLE
Add support to lock/unlock lba range

### PR DESCRIPTION
**Pull request by paulyoong**
The SPDK repo that this change points at still needs to be determined.

### CODE REVIEW

File: mayastor/src/core/descriptor.rs [hunk 1]
```diff
@@ -109,3 +165,41 @@ impl Debug for Descriptor {
         )
     }
 }
+
+extern "C" fn spdk_cb(
```
Comment (tiagolobocastro) [inline mayastor/src/core/descriptor.rs]: can we rename this to a more specific name, like spdk_range_callback or something like that?

File: mayastor/src/replicas/rebuild/rebuild_api.rs [hunk 2]
```diff
@@ -36,6 +36,26 @@ pub enum RebuildError {
     OpError { operation: String, state: String },
     #[snafu(display("Existing pending state {}", state,))]
     StatePending { state: String },
+    #[snafu(display(
+        "Failed to lock LBA range for blk {}, len {} ",
```
Comment (tiagolobocastro) [inline mayastor/src/replicas/rebuild/rebuild_api.rs]: did you mean to add the extra space at the end?

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 3]
```diff
@@ -205,6 +205,20 @@ impl RebuildJob {
             &mut self.tasks.buffers[id as usize]
         };
 
+        let mut nexus_desc = Bdev::open_by_name(&self.nexus, true).unwrap();
```
Comment (tiagolobocastro) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs]: should these unwraps be handled?

File: mayastor/tests/lock_lba_range.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        MayastorCliArgs,
+        MayastorEnvironment,
+        RangeContext,
+        Reactor,
+        Reactors,
+    },
+};
+use std::sync::{Arc, Mutex};
+
+const NEXUS_NAME: &str = "arbiter_nexus";
```
Comment (tiagolobocastro) [inline mayastor/tests/lock_lba_range.rs]: not using the arbiter now :)

File: mayastor/src/replicas/rebuild/rebuild_impl.rs [hunk 5]
```diff
@@ -205,6 +205,20 @@ impl RebuildJob {
...
+        let len = copy_buffer.len() as u64 / self.block_size;
+
+        let mut ctx =
+            RangeContext::new(blk, len, nexus_desc.get_channel().unwrap());
```
Comment (tiagolobocastro) [inline mayastor/src/replicas/rebuild/rebuild_impl.rs]: hmm is the channel going to be acquired and dropped once per segment? If so maybe we should cache it as part of the RebuildJob::new

File: mayastor/tests/lock_lba_range.rs [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+async fn create_nexus() {
+    let mut ch = Vec::new();
+    for i in 0 .. NUM_NEXUS_CHILDREN {
+        ch.push(get_dev(i));
+    }
+
+    match nexus_create(NEXUS_NAME, NEXUS_SIZE, None, &ch).await {
+        Ok(_) => println!("Created nexus"),
```
Comment (tiagolobocastro) [inline mayastor/tests/lock_lba_range.rs]: do we need these prints, maybe just call unwrap?

File: mayastor/tests/lock_lba_range.rs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    test_fini();
+}
+
+#[test]
+// Test locking a block range and then issuing a front-end I/O to an overlapping
+// range.
+// TODO: Add additional test for issuing front-end I/O then taking a lock
+fn lock_then_fe_io() {
+    test_ini();
+    {
```
Comment (tiagolobocastro) [inline mayastor/tests/lock_lba_range.rs:175]: do we need the new scope?

File: mayastor/tests/lock_lba_range.rs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+        // First lock is held, I/O should not succeed
+        assert!(io_receiver.try_recv().is_err());
+
+        // First unlock
+        let (s, r) = unbounded::<()>();
+        let ctx_clone = Arc::clone(&ctx);
+        reactor.send_future(async move {
+            unlock_range(&mut ctx_clone.lock().unwrap()).await;
+            trace!("*** RELEASE LOCK ***");
```
Comment (tiagolobocastro) [inline mayastor/tests/lock_lba_range.rs]: capitalization/wording is different to the previous test, should it be same/similar?

**Review by tiagolobocastro — COMMENTED** [review]
Looks good, the only thing of note is the IoChannel comment

File: mayastor/src/core/descriptor.rs (hunk 1 above)
Comment (paulyoong) [reply mayastor/src/core/descriptor.rs]: Done

File: mayastor/src/replicas/rebuild/rebuild_api.rs (hunk 2 above)
Comment (paulyoong) [reply mayastor/src/replicas/rebuild/rebuild_api.rs]: Changed the message

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 3 above)
Comment (paulyoong) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs]: Done

File: mayastor/tests/lock_lba_range.rs (hunk 4 above)
Comment (paulyoong) [reply mayastor/tests/lock_lba_range.rs]: Good catch, changed name

File: mayastor/src/replicas/rebuild/rebuild_impl.rs (hunk 5 above)
Comment (paulyoong) [reply mayastor/src/replicas/rebuild/rebuild_impl.rs]: Done

File: mayastor/tests/lock_lba_range.rs (hunk 6 above)
Comment (paulyoong) [reply mayastor/tests/lock_lba_range.rs]: Done

File: mayastor/tests/lock_lba_range.rs (hunk 7 above)
Comment (paulyoong) [reply mayastor/tests/lock_lba_range.rs:175]: If the new scope isn't used, the arc references used prevent the tests from cleaning up properly. Sticking the scope around it ensures the references are dropped before cleanup.

File: mayastor/tests/lock_lba_range.rs (hunk 8 above)
Comment (paulyoong) [reply mayastor/tests/lock_lba_range.rs]: Done

**Pull request closed without merging**